unified_planning/io/anml_writer.py

```diff
@@ -0,0 +1,329 @@
+    '''Expression converter to an ANML string.'''
+
+    DECIMAL_PRECISION = 10 # Number of decimal places to print real constants
+
+    def __init__(self, env: 'up.environment.Environment'):
+        walkers.DagWalker.__init__(self)
+        self.simplifier = walkers.Simplifier(env)
+
+    def convert(self, expression):
+        '''Converts the given expression to a ANML string.'''
+        return self.walk(self.simplifier.simplify(expression)) #NOTE maybe the converter could remove the first and last char, if they are '(' and ')'
+
+    def walk_exists(self, expression, args):
+        assert len(args) == 1
+        vars_string_list = [f'{str(cast(_UserType, v.type).name)} {v.name}' for v in expression.variables()]
+        return f'(exists({", ".join(vars_string_list)}) {{ {args[0]} }})' #NOTE i assumed vars are divided by ', '. #TODO check it
+
+    def walk_forall(self, expression, args):
+        assert len(args) == 1
+        vars_string_list = [f'{str(cast(_UserType, v.type).name)} {v.name}' for v in expression.variables()]
+        return f'(forall({", ".join(vars_string_list)}) {{ {args[0]} }})'
+
+    def walk_variable_exp(self, expression, args):
+        assert len(args) == 0
+        return expression.variable().name
+
+    def walk_and(self, expression, args):
+        assert len(args) > 1
+        return f'({" and ".join(args)})'
+
+    def walk_or(self, expression, args):
+        assert len(args) > 1
+        return f'({" or ".join(args)})'
+
+    def walk_not(self, expression, args):
+        assert len(args) == 1
+        return f'(not {args[0]})'
+
+    def walk_implies(self, expression, args):
+        assert len(args) == 2
+        return f'({args[0]} implies {args[1]})'
+
+    def walk_iff(self, expression, args):
+        assert len(args) == 2
+        return f'({args[0]} iff {args[1]})'
+
+    def walk_fluent_exp(self, expression, args):
+        fluent = expression.fluent()
+        if len(args) == 0:
+            return fluent.name
+        else:
+            return f'{fluent.name}({", ".join(args)})'
+
+    def walk_param_exp(self, expression, args):
+        assert len(args) == 0
+        return expression.parameter().name
+
+    def walk_object_exp(self, expression, args):
+        assert len(args) == 0
+        return expression.object().name
+
+    def walk_bool_constant(self, expression, args):
+        assert len(args) == 0
+        if expression.bool_constant_value():
+            return 'true'
+        return 'false'
+
+    def walk_real_constant(self, expression, args):#NOTE is this right? Same ANML Implementation
+        assert len(args) == 0
+        frac = expression.constant_value()
+
+        with localcontext() as ctx:
+            ctx.prec = self.DECIMAL_PRECISION
+            dec = frac.numerator / Decimal(frac.denominator, ctx)
+
+            if Fraction(dec) != frac:
+                warn("The ANML printer cannot exactly represent the real constant '%s'" % frac)
```

Comment:
real constants in the UP are Fraction

Reply:
In ANML you can write `frac.numerator/frac.denominator` and it should work I think